trio_websocket/__init__.py

```diff
@@ -153,6 +156,90 @@ def _url_to_host(url, ssl_context):
+    '''
+    wsproto = wsconnection.WSConnection(wsconnection.CLIENT, host=host,
+        resource=resource)
+    connection = WebSocketConnection(stream, wsproto, path=resource)
+    nursery.start_soon(connection._reader_task)
+    return connection
+
+
+def wrap_server_stream(nursery, stream):
+    '''
+    Wrap an arbitrary stream in a server-side ``WebSocketConnection``.
+
+    This is a low-level function only needed in rare cases. Most users should
+    call ``serve_websocket()`.
+
+    The returned connection object is not immediately usable: the caller should
+    wait for the ``open_handshake`` event before trying to use the connection.
+
+    :param nursery: A Trio nursery to run background tasks in.
+    :param stream: A Trio stream to be wrapped.
+    :rtype: WebSocketConnection
+    '''
+    wsproto = wsconnection.WSConnection(wsconnection.SERVER)
+    connection = WebSocketConnection(stream, wsproto)
+    nursery.start_soon(connection._reader_task)
+    return connection
+
+
+async def serve_websocket(handler, host, port, ssl_context,
+    handler_nursery=None, *, task_status=trio.TASK_STATUS_IGNORED):
+    '''
+    Serve a WebSocket over TCP.
+
+    This function supports the Trio nursery protocol, i.e.
+    ``server = await nursery.start(serve_websocket, …)``.
+
+    Note that if ``host`` is ``None`` and ``port`` is zero, then you may get
+    multiple listeners that have _different port numbers!_
+
+    :param coroutine handler: The coroutine called with the corresponding
+        WebSocketConnection on each new connection.  The call will be made
+        once the HTTP handshake completes, which notably implies that the
+        connection's `path` property will be valid.
+    :param host: The host interface to bind. This can be an address of an
+        interface, a name that resolves to an interface address (e.g.
+        ``localhost``), or a wildcard address like ``0.0.0.0`` for IPv4 or
+        ``::`` for IPv6. If ``None``, then all local interfaces are bound.
+    :type host: str, bytes, or None
+    :param int port: The port to bind to
+    :param ssl_context: The SSL context to use for encrypted connections, or
+        ``None`` for unencrypted connection.
+    :type ssl_context: SSLContext or None
+    :param handler_nursery: An optional nursery to spawn handlers and background
+        tasks in. If not specified, an internal nursery is used.
+    :param task_status: part of Trio nursery protocol
+    :returns: This function never returns unless cancelled.
+    '''
+    if ssl_context is None:
+        open_tcp_listeners = partial(trio.open_tcp_listeners, port, host=host)
+    else:
+        open_tcp_listeners = partial(trio.open_ssl_over_tcp_listeners, port,
+            ssl_context, host=host, https_compatible=True)
+    listeners = await open_tcp_listeners()
+    server = WebSocketServer(handler, listeners,
+        handler_nursery=handler_nursery)
+    await server.run(task_status=task_status)
```

Comment:
if user calls serve_websocket with start_soon instead of start, does he have a way to check if whatever this run intends to block on is complete?

the docstring above isn't clear on what's being ensured be using start() over start_soon()

Reply:
It's intended to wait for `trio.serve_listeners()` to finish starting up and also to return the server object, i.e. `server = nursery.start(serve_websocket, …)`. I think it would be an error to start this task with `start_soon`. I'll update the docstring. Not sure what else can be done here.